block/fork.go

```diff
@@ -0,0 +1,251 @@
+	if err != nil {
+		return err
+	}
+
+	if m.shouldStopNode(rollapp, lastBlock) {
+		err = m.createInstruction(rollapp)
+		if err != nil {
+			return err
+		}
+		m.freezeNode(ctx, rollapp, lastBlock, fmt.Errorf("fork update detected"))
+	}
+
+	return nil
+}
+
+func (m *Manager) createInstruction(rollapp *types.Rollapp) error {
+	obsoleteDrs, err := m.SLClient.GetObsoleteDrs()
+	if err != nil {
+		return err
+	}
+	instruction := types.Instruction{
+		Revision:            rollapp.Revision,
+		RevisionStartHeight: rollapp.RevisionStartHeight,
+		FaultyDRS:           obsoleteDrs,
+	}
+
+	err = types.PersistInstructionToDisk(m.RootDir, instruction)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// shouldStopNode determines if a rollapp node should be stopped based on revision criteria.
+//
+// This method checks two conditions to decide if a node should be stopped:
+// 1. If the next state height is greater than or equal to the rollapp's revision start height.
+// 2. If the block's app version (equivalent to revision) is less than the rollapp's revision
+func (m *Manager) shouldStopNode(rollapp *types.Rollapp, block *types.Block) bool {
+	revision := block.Header.Version.App
+	if m.State.NextHeight() >= rollapp.RevisionStartHeight && revision < rollapp.Revision {
+		return true
+	}
+
+	return false
+}
+
+// forkNeeded returns true if the fork file exists
+func (m *Manager) forkNeeded() (types.Instruction, bool) {
+	if instruction, err := types.LoadInstructionFromDisk(m.RootDir); err == nil {
+		return instruction, true
+	}
+
+	return types.Instruction{}, false
+}
+
+// handleSequencerForkTransition handles the sequencer fork transition
+func (m *Manager) handleSequencerForkTransition(instruction types.Instruction) {
+	consensusMsgs, err := m.prepareDRSUpgradeMessages(instruction.FaultyDRS)
+	if err != nil {
+		panic(fmt.Sprintf("prepare DRS upgrade messages: %v", err))
+	}
+	// Always bump the account sequences
+	consensusMsgs = append(consensusMsgs, &sequencers.MsgBumpAccountSequences{Authority: authtypes.NewModuleAddress("sequencers").String()})
+
+	err = m.handleCreationOfForkBlocks(instruction, consensusMsgs)
+	if err != nil {
+		panic(fmt.Sprintf("validate existing blocks: %v", err))
+	}
+
+	if err := m.handleForkBatchSubmission(instruction.RevisionStartHeight); err != nil {
+		panic(fmt.Sprintf("ensure batch exists: %v", err))
+	}
+}
+
+// prepareDRSUpgradeMessages prepares consensus messages for DRS upgrades.
+// It performs version validation and generates the necessary upgrade messages for the sequencer.
+//
+// The function implements the following logic:
+//   - If no faulty DRS version is provided (faultyDRS is nil), returns no messages
+//   - Validates the current DRS version against the potentially faulty version
+//   - Generates an upgrade message with the current valid DRS version
+func (m *Manager) prepareDRSUpgradeMessages(faultyDRS []uint32) ([]proto.Message, error) {
```

Comment:
the drs upgrade cnsensus message will be sent in all cases and the rdk will decide to plan an upgrade based on current rollapparams status, because this simplifies dymint flow and it is safer.

Reply:
I see thanks